common/src/main/java/io/crate/interval/NumericalIntervalParser.java

```diff
@@ -0,0 +1,103 @@
+
+
+class NumericalIntervalParser {
+
+    private NumericalIntervalParser() { }
+
+    static Period apply(String value) {
+        return apply(value, null, null);
+    }
+
+    static Period apply(String value, IntervalParser.Precision start, IntervalParser.Precision end) {
+        try {
+            return roundToPrecision(parseInteger(value), parseMilliSeconds(value), start, end);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Invalid interval format " + value);
+        }
+    }
+
+    private static Period roundToPrecision(int value,
+                                   int millis,
+                                   IntervalParser.Precision start,
+                                   IntervalParser.Precision end) {
+        if (start == null && end == null) {
+            return roundToSecondsWithMillis(value, millis);
+        }
+        if (start == IntervalParser.Precision.YEAR && end == null) {
+            return Period.years(value);
+        }
+        if (start == IntervalParser.Precision.YEAR && end == IntervalParser.Precision.MONTH) {
+            return Period.months(value);
+        }
+        if (start == IntervalParser.Precision.MONTH && end == null) {
+            return Period.months(value);
+        }
+        if (start == IntervalParser.Precision.DAY && end == null) {
+            return Period.days(value);
+        }
+        if (start == IntervalParser.Precision.DAY && end == IntervalParser.Precision.HOUR) {
+            return Period.hours(value);
+        }
+        if (start == IntervalParser.Precision.DAY && end == IntervalParser.Precision.MINUTE) {
+            return Period.minutes(value);
+        }
+        if (start == IntervalParser.Precision.DAY && end == IntervalParser.Precision.SECOND) {
+            return roundToSecondsWithMillis(value, millis);
+        }
+        if (start == IntervalParser.Precision.HOUR && end == null) {
+            return Period.hours(value);
+        }
+        if (start == IntervalParser.Precision.HOUR && end == IntervalParser.Precision.MINUTE) {
+            return Period.minutes(value);
+        }
+        if (start == IntervalParser.Precision.HOUR && end == IntervalParser.Precision.SECOND) {
+            return roundToSecondsWithMillis(value, millis);
+        }
+        if (start == IntervalParser.Precision.MINUTE && end == null) {
+            return Period.minutes(value);
+        }
+        if (start == IntervalParser.Precision.MINUTE && end == IntervalParser.Precision.SECOND) {
+            return Period.seconds(value);
+        }
+        if (start == IntervalParser.Precision.SECOND && end == null) {
+            return roundToSecondsWithMillis(value, millis);
+        }
+        throw new IllegalArgumentException(String.format("Invalid start %s and end %s combination", start, end));
+    }
+
+    private static Period roundToSecondsWithMillis(int seconds, int millis) {
```

Comment:
the `round` here seems a bit confusing. Isn't this keeping the full precision of both seconds & millis?